flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.calcite.rex.RexProgramBuilder;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Rule will merge Python {@link FlinkLogicalCalc} used in Map operation, Flatten {@link FlinkLogicalCalc}
+ * and Python {@link FlinkLogicalCalc} used in Map operation together.
+ */
+public class PythonCalcMergeRule extends RelOptRule {
+
+	public static final PythonCalcMergeRule INSTANCE = new PythonCalcMergeRule();
+
+	private PythonCalcMergeRule() {
+		super(operand(FlinkLogicalCalc.class,
+			operand(FlinkLogicalCalc.class,
+				operand(FlinkLogicalCalc.class, none()))),
+			"PythonCalcMergeRule");
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		FlinkLogicalCalc bottomCalc = call.rel(0);
+		FlinkLogicalCalc middleCalc = call.rel(1);
+		FlinkLogicalCalc topCalc = call.rel(2);
+		return isChainedRowBasedPythonFunction(bottomCalc, middleCalc, topCalc);
+	}
+
+	private boolean isChainedRowBasedPythonFunction(
+		FlinkLogicalCalc bottomCalc, FlinkLogicalCalc middleCalc, FlinkLogicalCalc topCalc) {
+		RexProgram bottomProgram = bottomCalc.getProgram();
+		List<RexNode> bottomProjects = bottomProgram.getProjectList()
+			.stream()
+			.map(bottomProgram::expandLocalRef)
+			.collect(Collectors.toList());
+
+		if (bottomProjects.size() != 1 || PythonUtil.isNonPythonCall(bottomProjects.get(0))) {
+			return false;
+		}
+
+		RexProgram topProgram = topCalc.getProgram();
+		List<RexNode> topProjects = topProgram.getProjectList()
+			.stream()
+			.map(topProgram::expandLocalRef)
+			.collect(Collectors.toList());
+		if (topProjects.size() != 1 || PythonUtil.isNonPythonCall(topProjects.get(0))) {
+			return false;
+		}
+
+		RexProgram middleProgram = middleCalc.getProgram();
+		List<RexNode> middleProjects = middleProgram.getProjectList()
+			.stream()
+			.map(middleProgram::expandLocalRef)
+			.collect(Collectors.toList());
+
+		return isInputsCorrespondWithUpstreamOutputs(bottomProjects, middleProjects) &&
+			isFlattenCalc(middleProjects);
+	}
+
+	private boolean isInputsCorrespondWithUpstreamOutputs(List<RexNode> bottomProjects, List<RexNode> middleProjects) {
+		RexCall pythonCall = (RexCall) bottomProjects.get(0);
+		List<RexNode> pythonCallInputs = pythonCall.getOperands();
+		if (pythonCallInputs.size() != middleProjects.size()) {
+			return false;
+		}
+		for (int i = 0; i < pythonCallInputs.size(); i++) {
+			RexNode input = pythonCallInputs.get(i);
+			if (input instanceof RexInputRef) {
+				if (((RexInputRef) input).getIndex() != i) {
+					return false;
+				}
+			} else {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	private boolean isFlattenCalc(List<RexNode> middleProjects) {
+		for (RexNode middleProject : middleProjects) {
+			if (middleProject instanceof RexFieldAccess) {
+				RexNode expr = ((RexFieldAccess) middleProject).getReferenceExpr();
+				if (expr instanceof RexInputRef) {
+					if (((RexInputRef) expr).getIndex() != 0) {
```

Comment:
This condition is too loose to determine whether the calc is a flatten calc.

Reply:
Yes. Good Catch.